File: app/src/main/java/io/ipoli/android/player/ExperienceForLevelGenerator.java
package io.ipoli.android.player;

import java.math.BigDecimal;
import java.math.BigInteger;

/**
 * Created by Venelin Valkov <venelin@curiousily.com>
 * on 5/30/16.
 */
public class ExperienceForLevelGenerator {

    private static BigDecimal fibonacci(int num) {
        BigDecimal phi = new BigDecimal((1 + Math.sqrt(5)) / 2);
        return phi.pow(num).subtract(BigDecimal.ONE.subtract(phi).pow(num)).divide(new BigDecimal(Math.sqrt(5)), BigDecimal.ROUND_HALF_UP);
    }

    public static BigInteger forLevel(int level) {
        if (level <= 1) {
            return BigInteger.ZERO;
        }
        return (fibonacci(level).add(BigDecimal.ONE)).toBigInteger().multiply(new BigInteger("10")).add(forLevel(level - 1));
    }
}

File: app/src/main/java/mypoli/android/player/ExperienceForLevelGenerator.kt
package mypoli.android.player

/**
 * Created by Venelin Valkov <venelin@mypoli.fun>
 * on 11/14/17.
 */
object ExperienceForLevelGenerator {

    private const val SWITCH_LEVEL = 6
    private const val LARGE_INCREASE = 80
    private const val SMALL_INCREASE = 50

    fun forLevel(level: Int): Long {

        if (level <= 1) {
            return 0
        }

        if (level == SWITCH_LEVEL) {
            val xpPrevLvl = forLevel(level - 1)
            val prevLevelDif = xpPrevLvl - forLevel(level - 2)
            return xpPrevLvl + prevLevelDif + LARGE_INCREASE * 2
        }

        val c = if (level < SWITCH_LEVEL) SMALL_INCREASE else LARGE_INCREASE

        return (c * (level - 1)) + forLevel(level - 1)
    }
}